firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,174 @@
+ * dereference each recorder at or before its Activity's onDestroy. Similar for Fragments.
+ */
+public class FrameMetricsRecorder {
+  private static final AndroidLogger logger = AndroidLogger.getInstance();
+
+  private final Activity activity;
+  private final FrameMetricsAggregator frameMetricsAggregator;
+  private final Map<Fragment, PerfFrameMetrics> subTraceMap;
+
+  private boolean isRecording = false;
+
+  /**
+   * Creates a recorder for a specific activity.
+   *
+   * @param activity the activity that the recorder is collecting data from.
+   */
+  public FrameMetricsRecorder(Activity activity) {
+    this(activity, new FrameMetricsAggregator(), new HashMap<>());
+  }
+
+  @VisibleForTesting
+  FrameMetricsRecorder(
+      Activity activity,
+      FrameMetricsAggregator frameMetricsAggregator,
+      Map<Fragment, PerfFrameMetrics> subTraceMap) {
+    this.activity = activity;
+    this.frameMetricsAggregator = frameMetricsAggregator;
+    this.subTraceMap = subTraceMap;
+  }
+
+  /** Starts recording FrameMetrics for the activity window. */
+  public void start() {
+    if (isRecording) {
+      logger.warn(
+          "FrameMetricsAggregator is already recording %s", activity.getClass().getSimpleName());
+      return;
+    }
+    frameMetricsAggregator.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window.
+   *
+   * @return FrameMetrics accumulated during the current recording.
+   */
+  public Optional<PerfFrameMetrics> stop() {
+    if (!isRecording) {
+      logger.warn("Cannot stop because no recording was started");
+      return Optional.absent();
+    }
+    if (!subTraceMap.isEmpty()) {
+      logger.warn(
+          "Sub-traces are still ongoing! Sub-traces should be stopped first before stopping Activity screen trace.");
+    }
+    Optional<PerfFrameMetrics> data = this.snapshot();
+    try {
+      frameMetricsAggregator.remove(activity);
+    } catch (IllegalArgumentException err) {
+      logger.debug(
+          "View not hardware accelerated. Unable to collect FrameMetrics. %s", err.toString());
+      return Optional.absent();
+    }
+    frameMetricsAggregator.reset();
+    isRecording = false;
+    return data;
+  }
+
+  /**
+   * Starts a sub-trace in the current recording.
+   *
+   * @param key a UI state to associate this sub-trace with. e.g.) fragment
+   */
+  public void startSubTrace(Fragment key) {
+    if (!isRecording) {
+      logger.warn("Cannot start sub-trace because FrameMetricsAggregator is not recording");
+      return;
+    }
+    if (subTraceMap.containsKey(key)) {
+      logger.warn(
+          "Cannot start sub-trace because one is already ongoing with the key %s",
+          key.getClass().getSimpleName());
+      return;
+    }
+    Optional<PerfFrameMetrics> snapshot = this.snapshot();
+    if (!snapshot.isAvailable()) {
+      logger.warn("startSubTrace(%s): snapshot() failed", key.getClass().getSimpleName());
```

Comment:
This information is not useful for the developers. Can we change this to something the developers can understand? Eg: `startFragment(%s): Unable to capture frame metrics when the fragment was shown on screen`.

Reply:
This message is meant to be shown to us, not to developers. In addition, `snapshot()` can be called in both `startFragment(...)` and `stopFragment(...)`, so I think it's more clear to just say "snapshot() failed".

Reply:
discussed offline and changing it to `logger.debug` would avoid developers seeing it, yet we still get the precise information that `snapshot()` failed.